Add ROOT-based event exporter

stognini opened the pull request:
This PR adds the option to export Celeritas' primaries to a ROOT file and read those primaries back into Celeritas.

- RootOffloadWriter writes data to a ROOT file using basic types to avoid the use of dictionaries.
- RootOffloadReader reads back the data in an event-by-event basis.

The reader can be used by a future `class RootPrimaryGenerator final : public G4VPrimaryGenerator { };`.

The use of a ROOT file considerably decreases the output file size, as expected. For the standard `celer-g4` test using `celeritas/app/data/ttbarbits-12evt-118prim.hepmc3`:

```
1.9M  simple-cms.offloaded.hepmc3
357K  simple-cms.offloaded.root
```

stognini committed:
Draft dump to root option to HepMC3PrimaryGenerator.cc

stognini committed:
Move functionality to a separate class

stognini committed:
Change to per-event tree entries

stognini committed:
Add HepMC3RootReader; WIP on other fronts

- Missing PDG from writer; EventReader has to be updated for that
- Revert back to 1 root entry == 1 primary
- celer-g4 has test code that has to be cleaned

stognini committed:
Update HepMC3RootReader to return events

stognini committed:
Reorg writer/reader

- Refactor class names
- Move to accel/detail
- Writer uses a vector<Primary> as input
- WIP: refactor reader to return vector<Primary> from ROOT

stognini committed:
Complete reader

stognini committed:
Add UPRootAutoSave custom deleter; Clean up code

stognini committed:
Add offloadRootOutputFile macro cmd option

stognini committed:
Merge branch 'develop' into dump-primaries

sethrj commented on src/accel/detail/RootOffloadReader.hh:
```diff
@@ -0,0 +1,111 @@
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file accel/detail/RootOffloadReader.hh
```
This should be `celeritas/io/EventReader` since we will need to use it from the celer-sim app.

sethrj commented on src/accel/detail/RootOffloadWriter.hh:
```diff
@@ -0,0 +1,91 @@
+ *
+ * One TTree entry represents one primary.
+ */
+class RootOffloadWriter
```
Likewise since this class doesn't depend on anything in `accel` and we can use it for writing out whatever primaries, put this in celeritas/io as `RootEventWriter`. The only thing that `OffloadWriter` did was to wrap a class for thread safety.

sethrj commented on src/accel/detail/RootOffloadWriter.cc:
```diff
@@ -0,0 +1,79 @@
+{
+    CELER_EXPECT(!primaries.empty());
+
+    std::scoped_lock{write_mutex_};
```
You should remove this. The `OffloadWriter` is the only thing that needs to be thread safe. If we need thread safety outside what ROOT provides we can use the existing OffloadWriter and make its `EventWriter` a generic `std::function<void(vector<Primary>)>`. Also note that the way the reader is structured, you **MUST** lock out the file when writing an event, otherwise primaries from different events could be interleaved.

stognini replied:
Just to be sure I got your point: so you're going to lock the `operator()` operation  itself, whether that's a HepMC3 or ROOT writer?

sethrj replied:
Right now the `OffloadWriter` in accel has a single purpose: to wrap an `EventWriter` with a thread safe lock. I think we'll use that same mechanism to wrap the RootWriter, since for the reading to work, we have to write all tracks consecutively for a single event.

sethrj commented on src/celeritas/ext/RootUniquePtr.root.cc:
```diff
@@ -29,6 +30,21 @@ void RootWritableDeleter<T>::operator()(T* ptr)
+    CELER_EXPECT(ptr);
+    CELER_LOG(debug) << "Writing " << ptr->ClassName() << " '"
+                     << ptr->GetName() << "'";
+    ptr->AutoSave();
```
Does any other interface have `AutoSave`  or just TTree? If the latter you should just make it a template specialization. Or just make separate deleters for ttree and tfile.

pcanal replied:
`AutoSave` is only seen in `TTree`.  `TObject->Write` should probably not be used in general except for `TFile`.   So it might actually indeed make sense to have 2 concrete deletor.  

Side note: Instead `TObject->Write`, especially in a threaded environment one would want to use `dir->WriteObject(ptr)`. i.e. require the writer to specify exactly where to write. (Only `TTree` and Histogram know which file they are attached to/intended for)

sethrj replied:
OK I agree we should have separate non-templated `RootTreeDeleter` and `RootFileDeleter` then.

sethrj commented on src/celeritas/ext/RootUniquePtr.root.cc:
```diff
@@ -48,6 +64,7 @@ void ExternDeleter<T>::operator()(T* ptr)
 //---------------------------------------------------------------------------//
 template struct RootWritableDeleter<TFile>;
 template struct RootWritableDeleter<TTree>;
+template struct RootAutoSaveDeleter<TTree>;
```
Delete the WritableDeleter<TTree> since it doesn't do the expected thing.

sethrj commented on src/accel/SharedParams.hh:
```diff
@@ -87,13 +88,20 @@ class SharedParams
+    {
+        return root_offload_writer_;
+    }
+
```
Delete all this code; we should be using the same offload writer interface: either HepMC3 or ROOT, we don't need separate copies of both. I just wanted you to add the `RootEvent{Reader,Writer}` class and I was going to add in a separate PR the extra bit of `std::functional` and "make a different writer based on the output extension" code needed to hook it in.

stognini committed:
Address comments

- Refactor RootOffload{Writer, Reader} -> RootEvent{Writer, Reader}
- Move to celeritas/io
- Remove root offload option from SharedParams/SetupOptionMessenger

stognini committed:
Use template specialization for UPRootWritable<TTree>

stognini committed:
Remove mutex from writer/reader

stognini committed:
Fix types

stognini committed:
Merge branch 'develop' into dump-primaries

sethrj commented on src/accel/CMakeLists.txt:
```diff
@@ -35,6 +35,10 @@ endif()
+
+  if (CELERITAS_USE_ROOT)
+    list(APPEND PRIVATE_DEPS ROOT::Tree)
+  endif()
```
This can be reverted since the ROOT code is no longer in `accel`.

sethrj replied:
@stognini You might have missed this comment earlier?

sethrj commented on src/accel/SetupOptions.hh:
```diff
@@ -112,6 +112,8 @@ struct SetupOptions
     //! Filename to dump a HepMC3 copy of offloaded tracks as events
     std::string offload_output_file;
+    //! Filename to dump a ROOT copy of offloaded tracks as events
+    std::string offload_root_output_file;
```
Remove this; we'll reuse `offload_output_file` and switch format based on the file extension.

sethrj replied:
Also this one?

stognini replied:
Ahh crap. Thanks for noticing this. I checked out back a few files from a previous commit just to test a `celer-g4` run and undid my fixes. I'll clean this up.

sethrj commented on src/celeritas/ext/RootUniquePtr.root.cc:
```diff
@@ -16,8 +16,7 @@ namespace celeritas
 /*!
- * Call `TObject->Write()` before deletion. Used by TFile and TTree writer
- * classes.
+ * Call `TObject->Write()` before deletion.
```
This would be a good place to mention that `Write` doesn't do what one would expect 😒

sethrj commented on src/celeritas/io/EventReader.cc:
```diff
@@ -86,21 +98,23 @@ auto EventReader::operator()() -> result_type
-
-        // Set the registered ID of the particle
-        primary.particle_id = particle_id;
-
```
I don't understand the use case for having an event reader that can't read the particle type. What can we do with primaries that are unknown particles? In all of the applications I can imagine, we'll be loading events into a simulation (or a test) that defines particle types already. I think the changes to EventReader.* should be reverted.

sethrj replied:
@stognini I think you missed this comment too.

sethrj commented on src/celeritas/io/RootEventReader.cc:
```diff
@@ -0,0 +1,101 @@
+                                 SPConstParticles params)
+    : params_(std::move(params))
+{
+    CELER_EXPECT(!filename.empty());
```
Add a `ScopedRootErrorHandler` here

sethrj commented on src/celeritas/io/RootEventReader.cc:
```diff
@@ -0,0 +1,101 @@
+ */
+auto RootEventReader::operator()() -> result_type
+{
+    CELER_EXPECT(entry_count_ <= num_entries_);
```
Add another `ScopedRootErrorHandler` here too.

sethrj commented on src/celeritas/io/RootEventReader.cc:
```diff
@@ -0,0 +1,101 @@
+        primaries.push_back(std::move(primary));
+    }
+
+    CELER_LOG_LOCAL(info) << "Read event " << this_evt_id << " with "
```
```suggestion
    CELER_LOG_LOCAL(debug) << "Read event " << this_evt_id << " with "
```

sethrj commented on src/celeritas/io/RootEventReader.cc:
```diff
@@ -0,0 +1,101 @@
+    return primaries;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Helper function to fetch leaves.
+ */
+template<class T>
+auto RootEventReader::from_leaf(char const* leaf_name) -> T
+{
+    CELER_EXPECT(ttree_);
+    auto const leaf = ttree_->GetLeaf(leaf_name);
+    CELER_ASSERT(leaf);
+    return static_cast<T>(leaf->GetValue());
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Helper function to fetch leaves containing `std::array<double, 3>`.
+ */
+Real3 RootEventReader::from_array_leaf(char const* leaf_name)
+{
+    CELER_EXPECT(ttree_);
+    auto const leaf = ttree_->GetLeaf(leaf_name);
+    CELER_ASSERT(leaf);
+    CELER_ASSERT(leaf->GetLen() == 3);
+    return {leaf->GetValue(0), leaf->GetValue(1), leaf->GetValue(2)};
+}
```
Depending on how often we write a non-dictionary ROOT file it might be h
```c++
class RootLeafReader
{
explicit RootLeafReader(TTree* tree);

template<class T>
void operator()(char const* leaf_name, T* value) { /* default implementation */ }

template<class T, class N>
void from_leaf(char const* leaf_name, Array<T, N>* value) { /* call GetValue on each item */ }

template<class T>
void from_leaf(char const* leaf_name, OpaqueId<T>* value) {...}
```

so that you can make a local `RootLeafReader read_leaf{ttree_.get()}; read_leaf("pos", p.pos);` etc.

Probably not worth it for this small file but something to keep in mind.

sethrj committed:
Refactor event IO test for reusability

sethrj committed:
Add root IO test

sethrj committed:
Prevent move/copy for event reader and writer

sethrj commented:
@stognini I refactored the event IO test (originally used for HepMC3 only) so that it can be reused for testing the ROOT reader and writer. Please take a look at the output: I think we're missing (1) the check to ensure that all primaries being written at once have the same event ID and (2) resetting the track IDs when reading back in. In fact I think we can entirely delete the track IDs from the ROOT output since we have to reset them sequentially anyway.

stognini committed:
Fix missing symbols error

stognini committed:
Update root unique ptrs

stognini committed:
Improve reader; Add ScopedRootErrorHandler; Remove TrackId branch

stognini committed:
Add scoped error handler; clean up

stognini commented:
> I refactored the event IO test (originally used for HepMC3 only) so that it can be reused for testing the ROOT reader and write

Thanks! 

>  In fact I think we can entirely delete the track IDs from the ROOT output since we have to reset them sequentially anyway.

Agreed. removed `track_id` branch from ROOT and manually set them during reading.

Now there is one detail that I am not sure how to proceed with the test: The `event_id` in `EventWriter` is set manually and sequentially, whereas the `RootEventWriter` initially simply assigned whatever `event_id` is stored in `celeritas::Primary` at the time, leading to the different results shown in the test. I have added an event id counter to the `RootEventWriter` to _manually match_ the exact behavior of `EventWriter`. Let me know if that is the way you prefer, or if I should amend the test and keep the `Root{reader/writer}` as previously (i.e. simply accepting that whatever information in `vector<Primary>` is correct to begin with).

stognini committed:
Update RootEventWriter to store contiguous event ids

stognini committed:
Fixup

- Update RootEventWriter constructor to use a SPRootFileManager, as the
  RootFileManager needs to be initialized or have a default constructor.
- Fix missing bits when CELERITAS_USE_ROOT=OFF

stognini committed:
Fixup; Minor improvements

sethrj commented:
> The event_id in EventWriter is set manually and sequentially, whereas the RootEventWriter initially simply assigned whatever event_id is stored in celeritas::Primary at the time, leading to the different results shown in the test. I have added an event id counter to the RootEventWriter to manually match the exact behavior of EventWriter. Let me know if that is the way you prefer, or if I should amend the test and keep the Root{reader/writer} as previously (i.e. simply accepting that whatever information in vector<Primary> is correct to begin with).

One big problem with the previous method is that interleaving the event IDs breaks the reader. Keeping the exact same behavior as the HepMC3 writer limits us, but it ensures no surprises when switching between `offload_output_file = "foo.hepmc3";` and `offload_output_file = "foo.root";`

stognini committed:
Merge branch 'develop' into dump-primaries

sethrj commented on src/celeritas/io/EventReader.hh:
```diff
@@ -46,13 +46,17 @@ class EventReader
 
     // Read a single event from the event record
     result_type operator()();
+    // Use a template and return T as HepMC3RootEvent or celeritas Primary?
```
This file should be reverted: event reader should always require particle params.

sethrj commented on src/celeritas/io/RootEventReader.cc:
```diff
@@ -0,0 +1,118 @@
+ */
+auto RootEventReader::operator()() -> result_type
+{
+    ScopedRootErrorHandler scoped_root_error;
+
+    CELER_EXPECT(entry_count_ <= num_entries_);
```
Put preconditions at the very very top. Same with the constructor above.

```suggestion
    CELER_EXPECT(entry_count_ <= num_entries_);

    ScopedRootErrorHandler scoped_root_error;
```

sethrj commented on src/celeritas/io/RootEventReader.hh:
```diff
@@ -0,0 +1,104 @@
+    CELER_ASSERT_UNREACHABLE();
+}
+
+template<class T>
+auto RootEventReader::from_leaf(char const*) -> T
+{
+    CELER_ASSERT_UNREACHABLE();
+}
+
+inline Real3 RootEventReader::from_array_leaf(char const*)
+{
+    CELER_ASSERT_UNREACHABLE();
+}
```
I don't think you need to provide definitions for these since they're never called if ROOT is disabled. Did you see a compiler warning-turned-error if they are gone?

stognini replied:
Interesting, they can indeed be removed. I haven't tested without them initially as I assumed that a missing definition would throw a Werror.

sethrj replied:
It's only a missing definition if it's called from somewhere.

sethrj committed:
Merge remote-tracking branch 'upstream/develop' into dump-primaries

stognini committed:
Revert EventReader; Address comments

sethrj reviewed (approved):
Looks good to me, sorry this took so long. I'll hook it up to `accel` next.
